src/server/frontend_wayland/wl_pointer.cpp

```diff
@@ -280,25 +296,33 @@ void mf::WlPointer::release()
     destroy_wayland_object();
 }
 
-WlStreamCursor::WlStreamCursor(
-    std::shared_ptr<mc::BufferStream> const& stream,
-    geom::Displacement hotspot) :
-    stream{stream},
-    hotspot{hotspot}
+WlSurfaceCursor::WlSurfaceCursor(mf::WlSurface* surface, geom::Displacement hotspot)
+    : surface{surface},
+      surface_destroyed{surface->destroyed_flag()},
+      stream{std::dynamic_pointer_cast<mc::BufferStream>(surface->stream)},
```

Comment:
We can and should just change `WlSurface::stream` to `std::shared_ptr<mir::compositor::BufferStream>`.